### PR TITLE
Switch to capnp encoding for PCBs.

### PR DESCRIPTION
**Pull request by kormat**
---

This change is [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion/760)

### CODE REVIEW

File: infrastructure/beacon_server/core.py [hunk 1]
```diff
@@ -144,7 +128,7 @@ def process_pcbs(self, pcbs, raw=True):
         for pcb in pcbs:
             if raw:
                 try:
-                    pcb = PathSegment(pcb)
+                    pcb = PathSegment.from_raw(pcb)
```
Comment (pszal) [inline infrastructure/beacon_server/core.py:131]: why we change convention?

Comment (kormat) [inline infrastructure/beacon_server/core.py:131]: The existing contention is bad, this is a saner replacement. Any time we need to pass in arguments to enable parsing it gets very messy. Using `from_raw` explicitly instead fixes this.

File: lib/packet/path.py [hunk 2]
```diff
@@ -555,12 +551,15 @@ def _copy_hofs(cls, ases, reverse=True):
...
-        for block in ases:
-            for ext in block.ext:
-                if ext.EXT_TYPE == MtuPcbExt.EXT_TYPE:
-                    mtu = min_mtu(mtu, ext.mtu)
-            hofs.append(copy.deepcopy(block.pcbm.hof))
+        mtu = float("inf")
+        for i, asm in enumerate(asms):
+            pcbm = asm.pcbm(0)
+            if i != 0:
+                # The upstream interface's mtu is irrelevant for the first
```
Comment (pszal) [inline lib/packet/path.py:569]: this mtu is per link, right?

File: lib/packet/path.py [hunk 3]
```diff
@@ -500,28 +502,23 @@ def _join_shortcuts(cls, up_segment, down_segment, point, peer=True):
...
-            ingress = pcbm.hof.ingress_if
-            if egress:
-                path.interfaces.append((pcbm.isd_as, egress))
-            if i != up_index:
-                path.interfaces.append((pcbm.isd_as, ingress))
+        for i, asm in enumerate(reversed(list(up_segment.iter_asms(up_index)))):
+            hof = asm.pcbm(0).hof()
+            if hof.egress_if:
+                path.interfaces.append((asm.isd_as(), hof.egress_if))
+            if i != 0:
```
Comment (pszal) [inline lib/packet/path.py]: `if i` (here and below) ?

File: lib/packet/path.py [hunk 4]
```diff
@@ -582,32 +581,35 @@ def _copy_segment_shortcut(cls, segment, index, up=True):
...
-                if (up_peer.isd_as == down_as.pcbm.isd_as and
-                        down_peer.isd_as == up_as.pcbm.isd_as):
-                    return up_peer.hof, down_peer.hof
+        up_ia = up_asm.isd_as()
+        down_ia = down_asm.isd_as()
+        for up_peer in up_asm.iter_pcbms(1):
+            up_hof = up_peer.hof()
+            for down_peer in down_asm.iter_pcbms(1):
+                down_hof = down_peer.hof()
+                if (up_peer.inIA() == down_ia and down_peer.inIA() == up_ia and
```
Comment (pszal) [inline lib/packet/path.py:620]: we need update our topo to test it ;-)

File: lib/packet/pcb.py [hunk 5]
```diff
@@ -88,457 +62,276 @@ def from_values(cls, isd_as, hof, ig_rev_token=None):  # pragma: no cover
...
+        d = self.p.to_dict()
+        d.setdefault('exts', []).append(ext)
+        self.p.from_dict(d)
+
+    def sig_pack(self, ver):
+        """
+        Pack for signing up for given version (defined by highest field number).
+        """
+        b = []
+        if ver >= 9:
```
Comment (pszal) [inline lib/packet/pcb.py:131]: what's that? Is 9 a constant?

File: lib/packet/pcb.py [hunk 6]
```diff
@@ -88,457 +62,276 @@ def from_values(cls, isd_as, hof, ig_rev_token=None):  # pragma: no cover
...
+    def iter_asms(self, start=0):
+        for i in range(start, len(self.p.asms)):
+            yield self.asm(i)
+
+    def is_sibra(self):
+        return bool(self.p.exts.sibra.id)
+
+    def sig_pack(self, ver=3):
+        b = []
+        if ver >= 3:
```
Comment (pszal) [inline lib/packet/pcb.py:199]: again?

**Pull request merged**